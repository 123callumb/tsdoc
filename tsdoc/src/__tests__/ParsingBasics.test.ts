import {
  CoreModifierTagSet,
  DocComment,
  ParserContext
} from '../index';
import { TestHelpers } from '../parser/__tests__/TestHelpers';

test('01 Simple @beta and @internal extraction', () => {
  const parserContext: ParserContext = TestHelpers.parseAndMatchDocCommentSnapshot([
    '/**',
    ' * START @beta',
    ' * @unknownTag',
    ' * @internal @internal END',
    ' */'
  ].join('\n'));

  const docComment: DocComment = parserContext.docComment;
  const modifierTagSet: CoreModifierTagSet = docComment.modifierTagSet;

  expect(modifierTagSet.isAlpha).toEqual(false);
  expect(modifierTagSet.isBeta).toEqual(true);
  expect(modifierTagSet.isInternal).toEqual(true);
});
