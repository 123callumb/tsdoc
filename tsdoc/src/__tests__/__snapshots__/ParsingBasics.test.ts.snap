// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`01 Simple @beta and @internal extraction 1`] = `
Object {
  "_0_lines": Array [
    "START @beta",
    "@unknownTag",
    "@internal @internal END",
  ],
  "_1_gaps": Array [
    "@internal",
  ],
  "_2_summarySection": Object {
    "kind": "Section",
    "nodes": Array [
      Object {
        "kind": "Paragraph",
        "nodes": Array [
          Object {
            "kind": "PlainText",
            "nodeExcerpt": "START ",
          },
          Object {
            "kind": "SoftBreak",
            "nodeExcerpt": "[n]",
          },
          Object {
            "kind": "BlockTag",
            "nodeExcerpt": "@unknownTag",
          },
          Object {
            "kind": "SoftBreak",
            "nodeExcerpt": "[n]",
          },
          Object {
            "kind": "PlainText",
            "nodeExcerpt": " ",
          },
          Object {
            "kind": "PlainText",
            "nodeExcerpt": " END",
          },
          Object {
            "kind": "SoftBreak",
            "nodeExcerpt": "[n]",
          },
        ],
      },
    ],
  },
  "_3_remarksBlock": undefined,
  "_4_customBlocks": Array [],
  "_5_paramBlocks": Array [],
  "_6_returnsBlock": undefined,
  "_7_modifierTags": Array [
    Object {
      "kind": "BlockTag",
      "nodeExcerpt": "@beta",
    },
    Object {
      "kind": "BlockTag",
      "nodeExcerpt": "@internal",
    },
  ],
  "_8_logMessages": Array [],
}
`;

exports[`02 A basic TSDoc comment with all components 1`] = `
Object {
  "_0_lines": Array [
    "Returns the average of two numbers.",
    "",
    "@remarks",
    "This method is part of the {@link core-library#Statistics | Statistics subsystem}.",
    "",
    "@customBlock",
    "This is a custom block containing an @undefinedBlockTag",
    "",
    "@param x - The first input number",
    "@param y - The second input number",
    "@returns The arithmetic mean of [c]x[c] and [c]y[c]",
    "",
    "@beta @customModifier",
  ],
  "_1_gaps": Array [],
  "_2_summarySection": Object {
    "kind": "Section",
    "nodes": Array [
      Object {
        "kind": "Paragraph",
        "nodes": Array [
          Object {
            "kind": "PlainText",
            "nodeExcerpt": "Returns the average of two numbers.",
          },
          Object {
            "kind": "SoftBreak",
            "nodeExcerpt": "[n]",
          },
          Object {
            "kind": "SoftBreak",
            "nodeExcerpt": "[n]",
          },
        ],
      },
    ],
  },
  "_3_remarksBlock": Object {
    "kind": "Block",
    "nodes": Array [
      Object {
        "kind": "BlockTag",
        "nodeExcerpt": "@remarks",
      },
      Object {
        "kind": "Paragraph",
        "nodes": Array [
          Object {
            "kind": "SoftBreak",
            "nodeExcerpt": "[n]",
          },
          Object {
            "kind": "PlainText",
            "nodeExcerpt": "This method is part of the ",
          },
          Object {
            "kind": "LinkTag",
            "nodes": Array [
              Object {
                "kind": "Particle: openingDelimiter",
                "nodeExcerpt": "{",
              },
              Object {
                "kind": "Particle: tagName",
                "nodeExcerpt": "@link",
                "nodeSpacing": " ",
              },
              Object {
                "kind": "DeclarationReference",
                "nodes": Array [
                  Object {
                    "kind": "Particle: packageName",
                    "nodeExcerpt": "core-library",
                  },
                  Object {
                    "kind": "Particle: importHash",
                    "nodeExcerpt": "#",
                  },
                  Object {
                    "kind": "MemberReference",
                    "nodes": Array [
                      Object {
                        "kind": "MemberIdentifier",
                        "nodes": Array [
                          Object {
                            "kind": "Particle: identifier",
                            "nodeExcerpt": "Statistics",
                            "nodeSpacing": " ",
                          },
                        ],
                      },
                    ],
                  },
                ],
              },
              Object {
                "kind": "Particle: pipe",
                "nodeExcerpt": "|",
              },
              Object {
                "kind": "Particle: linkText",
                "nodeExcerpt": " Statistics subsystem",
              },
              Object {
                "kind": "Particle: closingDelimiter",
                "nodeExcerpt": "}",
              },
            ],
          },
          Object {
            "kind": "PlainText",
            "nodeExcerpt": ".",
          },
          Object {
            "kind": "SoftBreak",
            "nodeExcerpt": "[n]",
          },
          Object {
            "kind": "SoftBreak",
            "nodeExcerpt": "[n]",
          },
        ],
      },
    ],
  },
  "_4_customBlocks": Array [
    Object {
      "kind": "Block",
      "nodes": Array [
        Object {
          "kind": "BlockTag",
          "nodeExcerpt": "@customBlock",
        },
        Object {
          "kind": "Paragraph",
          "nodes": Array [
            Object {
              "kind": "SoftBreak",
              "nodeExcerpt": "[n]",
            },
            Object {
              "kind": "PlainText",
              "nodeExcerpt": "This is a custom block containing an ",
            },
            Object {
              "kind": "BlockTag",
              "nodeExcerpt": "@undefinedBlockTag",
            },
            Object {
              "kind": "SoftBreak",
              "nodeExcerpt": "[n]",
            },
            Object {
              "kind": "SoftBreak",
              "nodeExcerpt": "[n]",
            },
          ],
        },
      ],
    },
  ],
  "_5_paramBlocks": Array [
    Object {
      "kind": "ParamBlock",
      "nodes": Array [
        Object {
          "kind": "BlockTag",
          "nodeExcerpt": "@param",
          "nodeSpacing": " ",
        },
        Object {
          "kind": "Particle: parameterName",
          "nodeExcerpt": "x",
          "nodeSpacing": " ",
        },
        Object {
          "kind": "Particle: hyphen",
          "nodeExcerpt": "-",
          "nodeSpacing": " ",
        },
        Object {
          "kind": "Paragraph",
          "nodes": Array [
            Object {
              "kind": "PlainText",
              "nodeExcerpt": "The first input number",
            },
            Object {
              "kind": "SoftBreak",
              "nodeExcerpt": "[n]",
            },
          ],
        },
      ],
    },
    Object {
      "kind": "ParamBlock",
      "nodes": Array [
        Object {
          "kind": "BlockTag",
          "nodeExcerpt": "@param",
          "nodeSpacing": " ",
        },
        Object {
          "kind": "Particle: parameterName",
          "nodeExcerpt": "y",
          "nodeSpacing": " ",
        },
        Object {
          "kind": "Particle: hyphen",
          "nodeExcerpt": "-",
          "nodeSpacing": " ",
        },
        Object {
          "kind": "Paragraph",
          "nodes": Array [
            Object {
              "kind": "PlainText",
              "nodeExcerpt": "The second input number",
            },
            Object {
              "kind": "SoftBreak",
              "nodeExcerpt": "[n]",
            },
          ],
        },
      ],
    },
  ],
  "_6_returnsBlock": Object {
    "kind": "Block",
    "nodes": Array [
      Object {
        "kind": "BlockTag",
        "nodeExcerpt": "@returns",
      },
      Object {
        "kind": "Paragraph",
        "nodes": Array [
          Object {
            "kind": "PlainText",
            "nodeExcerpt": " The arithmetic mean of ",
          },
          Object {
            "kind": "CodeSpan",
            "nodes": Array [
              Object {
                "kind": "Particle: openingDelimiter",
                "nodeExcerpt": "[c]",
              },
              Object {
                "kind": "Particle: code",
                "nodeExcerpt": "x",
              },
              Object {
                "kind": "Particle: closingDelimiter",
                "nodeExcerpt": "[c]",
              },
            ],
          },
          Object {
            "kind": "PlainText",
            "nodeExcerpt": " and ",
          },
          Object {
            "kind": "CodeSpan",
            "nodes": Array [
              Object {
                "kind": "Particle: openingDelimiter",
                "nodeExcerpt": "[c]",
              },
              Object {
                "kind": "Particle: code",
                "nodeExcerpt": "y",
              },
              Object {
                "kind": "Particle: closingDelimiter",
                "nodeExcerpt": "[c]",
              },
            ],
          },
          Object {
            "kind": "SoftBreak",
            "nodeExcerpt": "[n]",
          },
          Object {
            "kind": "SoftBreak",
            "nodeExcerpt": "[n]",
          },
          Object {
            "kind": "PlainText",
            "nodeExcerpt": " ",
          },
          Object {
            "kind": "SoftBreak",
            "nodeExcerpt": "[n]",
          },
        ],
      },
    ],
  },
  "_7_modifierTags": Array [
    Object {
      "kind": "BlockTag",
      "nodeExcerpt": "@beta",
    },
    Object {
      "kind": "BlockTag",
      "nodeExcerpt": "@customModifier",
    },
  ],
  "_8_logMessages": Array [],
}
`;

exports[`03 Incomplete @param blocks 1`] = `
Object {
  "_0_lines": Array [
    "@param - The first input number",
    "@param y The second input number",
    "@returns The arithmetic mean of [c]x[c] and [c]y[c]",
  ],
  "_1_gaps": Array [],
  "_2_summarySection": Object {
    "kind": "Section",
  },
  "_3_remarksBlock": undefined,
  "_4_customBlocks": Array [],
  "_5_paramBlocks": Array [
    Object {
      "kind": "ParamBlock",
      "nodes": Array [
        Object {
          "kind": "BlockTag",
          "nodeExcerpt": "@param",
        },
        Object {
          "kind": "Particle: parameterName",
        },
        Object {
          "kind": "Particle: hyphen",
        },
        Object {
          "kind": "Paragraph",
          "nodes": Array [
            Object {
              "kind": "PlainText",
              "nodeExcerpt": " - The first input number",
            },
            Object {
              "kind": "SoftBreak",
              "nodeExcerpt": "[n]",
            },
          ],
        },
      ],
    },
    Object {
      "kind": "ParamBlock",
      "nodes": Array [
        Object {
          "kind": "BlockTag",
          "nodeExcerpt": "@param",
        },
        Object {
          "kind": "Particle: parameterName",
        },
        Object {
          "kind": "Particle: hyphen",
        },
        Object {
          "kind": "Paragraph",
          "nodes": Array [
            Object {
              "kind": "PlainText",
              "nodeExcerpt": " y The second input number",
            },
            Object {
              "kind": "SoftBreak",
              "nodeExcerpt": "[n]",
            },
          ],
        },
      ],
    },
  ],
  "_6_returnsBlock": Object {
    "kind": "Block",
    "nodes": Array [
      Object {
        "kind": "BlockTag",
        "nodeExcerpt": "@returns",
      },
      Object {
        "kind": "Paragraph",
        "nodes": Array [
          Object {
            "kind": "PlainText",
            "nodeExcerpt": " The arithmetic mean of ",
          },
          Object {
            "kind": "CodeSpan",
            "nodes": Array [
              Object {
                "kind": "Particle: openingDelimiter",
                "nodeExcerpt": "[c]",
              },
              Object {
                "kind": "Particle: code",
                "nodeExcerpt": "x",
              },
              Object {
                "kind": "Particle: closingDelimiter",
                "nodeExcerpt": "[c]",
              },
            ],
          },
          Object {
            "kind": "PlainText",
            "nodeExcerpt": " and ",
          },
          Object {
            "kind": "CodeSpan",
            "nodes": Array [
              Object {
                "kind": "Particle: openingDelimiter",
                "nodeExcerpt": "[c]",
              },
              Object {
                "kind": "Particle: code",
                "nodeExcerpt": "y",
              },
              Object {
                "kind": "Particle: closingDelimiter",
                "nodeExcerpt": "[c]",
              },
            ],
          },
          Object {
            "kind": "SoftBreak",
            "nodeExcerpt": "[n]",
          },
        ],
      },
    ],
  },
  "_7_modifierTags": Array [],
  "_8_logMessages": Array [
    "(2,4): The @param block should be followed by a parameter name",
    "(3,4): The @param block should be followed by a parameter name and then a hyphen",
  ],
}
`;

exports[`03 Jumbled order 1`] = `
Object {
  "_0_lines": Array [
    "Returns the average of two numbers. @remarks This method is part of the",
    "{@link core-library#Statistics | Statistics subsystem}.",
    "@beta @customModifier",
    "@returns The arithmetic mean of [c]x[c] and [c]y[c]",
    "@param x - The first input number @param y - The second input number",
    "@customBlock",
    "This is a custom block containing an @undefinedBlockTag",
  ],
  "_1_gaps": Array [],
  "_2_summarySection": Object {
    "kind": "Section",
    "nodes": Array [
      Object {
        "kind": "Paragraph",
        "nodes": Array [
          Object {
            "kind": "PlainText",
            "nodeExcerpt": "Returns the average of two numbers. ",
          },
        ],
      },
    ],
  },
  "_3_remarksBlock": Object {
    "kind": "Block",
    "nodes": Array [
      Object {
        "kind": "BlockTag",
        "nodeExcerpt": "@remarks",
      },
      Object {
        "kind": "Paragraph",
        "nodes": Array [
          Object {
            "kind": "PlainText",
            "nodeExcerpt": " This method is part of the",
          },
          Object {
            "kind": "SoftBreak",
            "nodeExcerpt": "[n]",
          },
          Object {
            "kind": "LinkTag",
            "nodes": Array [
              Object {
                "kind": "Particle: openingDelimiter",
                "nodeExcerpt": "{",
              },
              Object {
                "kind": "Particle: tagName",
                "nodeExcerpt": "@link",
                "nodeSpacing": " ",
              },
              Object {
                "kind": "DeclarationReference",
                "nodes": Array [
                  Object {
                    "kind": "Particle: packageName",
                    "nodeExcerpt": "core-library",
                  },
                  Object {
                    "kind": "Particle: importHash",
                    "nodeExcerpt": "#",
                  },
                  Object {
                    "kind": "MemberReference",
                    "nodes": Array [
                      Object {
                        "kind": "MemberIdentifier",
                        "nodes": Array [
                          Object {
                            "kind": "Particle: identifier",
                            "nodeExcerpt": "Statistics",
                            "nodeSpacing": " ",
                          },
                        ],
                      },
                    ],
                  },
                ],
              },
              Object {
                "kind": "Particle: pipe",
                "nodeExcerpt": "|",
              },
              Object {
                "kind": "Particle: linkText",
                "nodeExcerpt": " Statistics subsystem",
              },
              Object {
                "kind": "Particle: closingDelimiter",
                "nodeExcerpt": "}",
              },
            ],
          },
          Object {
            "kind": "PlainText",
            "nodeExcerpt": ".",
          },
          Object {
            "kind": "SoftBreak",
            "nodeExcerpt": "[n]",
          },
          Object {
            "kind": "PlainText",
            "nodeExcerpt": " ",
          },
          Object {
            "kind": "SoftBreak",
            "nodeExcerpt": "[n]",
          },
        ],
      },
    ],
  },
  "_4_customBlocks": Array [
    Object {
      "kind": "Block",
      "nodes": Array [
        Object {
          "kind": "BlockTag",
          "nodeExcerpt": "@customBlock",
        },
        Object {
          "kind": "Paragraph",
          "nodes": Array [
            Object {
              "kind": "SoftBreak",
              "nodeExcerpt": "[n]",
            },
            Object {
              "kind": "PlainText",
              "nodeExcerpt": "This is a custom block containing an ",
            },
            Object {
              "kind": "BlockTag",
              "nodeExcerpt": "@undefinedBlockTag",
            },
            Object {
              "kind": "SoftBreak",
              "nodeExcerpt": "[n]",
            },
          ],
        },
      ],
    },
  ],
  "_5_paramBlocks": Array [
    Object {
      "kind": "ParamBlock",
      "nodes": Array [
        Object {
          "kind": "BlockTag",
          "nodeExcerpt": "@param",
          "nodeSpacing": " ",
        },
        Object {
          "kind": "Particle: parameterName",
          "nodeExcerpt": "x",
          "nodeSpacing": " ",
        },
        Object {
          "kind": "Particle: hyphen",
          "nodeExcerpt": "-",
          "nodeSpacing": " ",
        },
        Object {
          "kind": "Paragraph",
          "nodes": Array [
            Object {
              "kind": "PlainText",
              "nodeExcerpt": "The first input number ",
            },
          ],
        },
      ],
    },
    Object {
      "kind": "ParamBlock",
      "nodes": Array [
        Object {
          "kind": "BlockTag",
          "nodeExcerpt": "@param",
          "nodeSpacing": " ",
        },
        Object {
          "kind": "Particle: parameterName",
          "nodeExcerpt": "y",
          "nodeSpacing": " ",
        },
        Object {
          "kind": "Particle: hyphen",
          "nodeExcerpt": "-",
          "nodeSpacing": " ",
        },
        Object {
          "kind": "Paragraph",
          "nodes": Array [
            Object {
              "kind": "PlainText",
              "nodeExcerpt": "The second input number",
            },
            Object {
              "kind": "SoftBreak",
              "nodeExcerpt": "[n]",
            },
          ],
        },
      ],
    },
  ],
  "_6_returnsBlock": Object {
    "kind": "Block",
    "nodes": Array [
      Object {
        "kind": "BlockTag",
        "nodeExcerpt": "@returns",
      },
      Object {
        "kind": "Paragraph",
        "nodes": Array [
          Object {
            "kind": "PlainText",
            "nodeExcerpt": " The arithmetic mean of ",
          },
          Object {
            "kind": "CodeSpan",
            "nodes": Array [
              Object {
                "kind": "Particle: openingDelimiter",
                "nodeExcerpt": "[c]",
              },
              Object {
                "kind": "Particle: code",
                "nodeExcerpt": "x",
              },
              Object {
                "kind": "Particle: closingDelimiter",
                "nodeExcerpt": "[c]",
              },
            ],
          },
          Object {
            "kind": "PlainText",
            "nodeExcerpt": " and ",
          },
          Object {
            "kind": "CodeSpan",
            "nodes": Array [
              Object {
                "kind": "Particle: openingDelimiter",
                "nodeExcerpt": "[c]",
              },
              Object {
                "kind": "Particle: code",
                "nodeExcerpt": "y",
              },
              Object {
                "kind": "Particle: closingDelimiter",
                "nodeExcerpt": "[c]",
              },
            ],
          },
          Object {
            "kind": "SoftBreak",
            "nodeExcerpt": "[n]",
          },
        ],
      },
    ],
  },
  "_7_modifierTags": Array [
    Object {
      "kind": "BlockTag",
      "nodeExcerpt": "@beta",
    },
    Object {
      "kind": "BlockTag",
      "nodeExcerpt": "@customModifier",
    },
  ],
  "_8_logMessages": Array [],
}
`;
